solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -766,16 +766,29 @@ public Document doc(int i, Set<String> fields) throws IOException {
     }
 
     final DirectoryReader reader = getIndexReader();
-    if (!enableLazyFieldLoading || fields == null) {
-      d = reader.document(i);
+    boolean containsAllFields = true;
+    
+    if (fields != null) {
+      if (enableLazyFieldLoading) {
+        final SetNonLazyFieldSelector visitor = new SetNonLazyFieldSelector(fields, reader, i);
+        reader.document(i, visitor);
+        d = visitor.doc;
+      } else {
+        d = reader.document(i, fields);
+        containsAllFields = false;
+      }
     } else {
-      final SetNonLazyFieldSelector visitor = new SetNonLazyFieldSelector(fields, reader, i);
-      reader.document(i, visitor);
-      d = visitor.doc;
+      d = reader.document(i);
     }
 
     if (documentCache != null) {
-      documentCache.put(i, d);
+      // Only cache the already retrieved document if it is complete... 
+      if (containsAllFields) {
+        documentCache.put(i, d);
+      } else {
+        // ...and retrieve a complete document for caching otherwise.
+        documentCache.put(i, reader.document(i));
+      }
```

Comment:
@shalinmangar I've made a pass at restoring the previous caching behavior (with some comments around the rationale). It feels like we might have a complete solution at this point.